src/init/features/dataconnect/sdk.ts

```diff
@@ -109,103 +105,83 @@ async function askQuestions(setup: Setup, config: Config): Promise<SDKInfo> {
     choices: connectorChoices,
   });
 
-  const newConnectorYaml = JSON.parse(JSON.stringify(connectorInfo.connectorYaml)) as ConnectorYaml;
-  if (!newConnectorYaml.generate) {
-    newConnectorYaml.generate = {};
+  const connectorYaml = JSON.parse(JSON.stringify(connectorInfo.connectorYaml)) as ConnectorYaml;
+  const newConnectorYaml = generateSdkYaml(
+    targetPlatform,
+    connectorYaml,
+    connectorInfo.directory,
+    appDir,
+  );
+
+  const shouldGenerate = !!(
+    setup.projectId &&
+    (await confirm({
+      message: "Would you like to generate SDK code now?",
+      default: true,
+    }))
+  );
+  // TODO: Prompt user  about adding generated paths to .gitignore
+  const connectorYamlContents = yaml.stringify(newConnectorYaml);
+  connectorInfo.connectorYaml = newConnectorYaml;
+  const displayIOSWarning = targetPlatform === Platform.IOS;
+  return { connectorYamlContents, connectorInfo, shouldGenerate, displayIOSWarning };
+}
+
+export function generateSdkYaml(
+  targetPlatform: Platform,
+  connectorYaml: ConnectorYaml,
+  connectorDir: string, // path.relative expe cts folder as first arg
+  appDir: string,
+): ConnectorYaml {
+  if (!connectorYaml.generate) {
+    connectorYaml.generate = {};
   }
 
-  let displayIOSWarning = false;
   if (targetPlatform === Platform.IOS) {
-    const outputDir =
-      newConnectorYaml.generate.swiftSdk?.outputDir ||
-      path.relative(connectorInfo.directory, path.join(appDir, `generated/swift`));
-    const pkg =
-      newConnectorYaml.generate.swiftSdk?.package ??
-      upperFirst(camelCase(newConnectorYaml.connectorId));
-    const swiftSdk = { outputDir, package: pkg };
-    newConnectorYaml.generate.swiftSdk = swiftSdk;
-    displayIOSWarning = true;
+    const swiftSdk = {
+      outputDir: path.relative(connectorDir, path.join(appDir, `dataconnect-generated/swift`)),
+      package: upperFirst(camelCase(connectorYaml.connectorId)),
+    };
+    connectorYaml.generate.swiftSdk = swiftSdk;
   }
 
   if (targetPlatform === Platform.WEB) {
-    const outputDir =
-      newConnectorYaml.generate.javascriptSdk?.outputDir ||
-      path.relative(
-        connectorInfo.directory,
-        path.join(appDir, `generated/javascript/${newConnectorYaml.connectorId}`),
-      );
-    const pkg =
-      newConnectorYaml.generate.javascriptSdk?.package ??
-      `@firebasegen/${connectorInfo.connectorYaml.connectorId}`;
-
     const javascriptSdk: JavascriptSDK = {
-      outputDir,
-      package: pkg,
-      packageJsonDir: newConnectorYaml.generate.javascriptSdk?.packageJsonDir,
+      outputDir: path.relative(connectorDir, path.join(appDir, `dataconnect-generated/js`)),
+      package: `@firebasegen/${connectorYaml.connectorId}`,
+      // If appDir has package.json, this would install JS SDK there.
+      // Otherwise, emulator would ignore it. Always add it here in case `package.json` is added later.
+      packageJsonDir: path.relative(connectorDir, appDir),
```

Comment:
https://github.com/firebase/firebase-tools/pull/7777#discussion_r1783526292

Maneesh brought up a good point about missing `package.json`.

I don't think there is any harm to always set it. Emulator can ignore it and print a more clear debug log: http://cl/682785528

Reply:
`      // If appDir has package.json, Emulator would install JS SDK for every reload.` - this is not accurate. We only add the dependency if it doesn't already exist in the package.json. So it would add the first time, and then skip the rest.

Also, fwiw, this doesn't automatically install the JS SDK, it automatically installs your generated SDK package, and your package manager will automatically install the JS SDK depending on which one you're using.

What's the reason to remove the prompt? It's an optional field, so I feel like we might want to just keep the prompt.

Reply:
This is fine with me